custom_components/unifiprotect/sensor.py

```diff
@@ -157,25 +298,62 @@ def __init__(
         self,
         protect: ProtectApiClient,
         protect_data: UnifiProtectData,
-        device: ProtectAdoptableDeviceModel,
+        device: ProtectDeviceModel,
         description: UnifiProtectSensorEntityDescription,
     ):
         """Initialize an Unifi Protect sensor."""
         self.entity_description: UnifiProtectSensorEntityDescription = description
         super().__init__(protect, protect_data, device, description)
         self._attr_name = f"{self.device.name} {self.entity_description.name}"
 
-    @property
-    def native_value(self) -> StateType | date | datetime:
-        """Return the state of the sensor."""
-        value = get_nested_attr(self.device, self.entity_description.ufp_value)
-
-        assert isinstance(value, (str, int, float, datetime)) or value is None
-        if isinstance(value, datetime) and not above_ha_version(2021, 12):
-            return value.replace(microsecond=0).isoformat()
+    @callback
+    def _async_nvr_value(self) -> float:
+        assert isinstance(self.device, NVR)
+        if self.entity_description.key == _KEY_MEMORY:
+            memory = self.device.system_info.memory
+            value = (1 - memory.available / memory.total) * 100
+        else:
+            record_type, subtype = self.entity_description.key.split("_")
+            if record_type == "record":
+                record_type_attr = "recording_type_distributions"
+                subtype_attr = "recording_type"
+            else:
+                record_type_attr = "resolution_distributions"
+                subtype_attr = "resolution"
 
+            distributions = self.device.storage_stats.storage_distribution
+            for item in getattr(distributions, record_type_attr):
+                if getattr(item, subtype_attr) == subtype:
+                    value = getattr(item, "percentage")
+                    break
         return value
 
+    @callback
+    def _async_update_device_from_protect(self) -> None:
+        super()._async_update_device_from_protect()
+
+        if self.entity_description.ufp_value is None:
+            new_value = self._async_nvr_value()
+        else:
+            new_value = get_nested_attr(self.device, self.entity_description.ufp_value)
+
+        if isinstance(new_value, timedelta):
+            new_value = new_value.total_seconds()
+
+        if isinstance(new_value, datetime):
+            # protection to prevent uptime from changing if UniFi Protect changes values slightly
+            if hasattr(self, "_attr_native_value") and self._attr_native_value:
+                diff = abs((self._attr_native_value - new_value).total_seconds())
+                if diff < 5:
+                    return
+
+            if not above_ha_version(2021, 12):
+                new_value = new_value.isoformat()
```

Comment:
Thoughts on this? I changed it to use `_attr_native_value` so I can see the previous value and then require the diff to be more then 5 seconds before updating. My G4 on master kept getting new values even though I had `replace(microsecond=0)` so that must mean the second value was slightly changing over time as well. It was consistently Decmember 3rd at 10:26 PM though.